### PR TITLE
Add modular applicability

### PR DESCRIPTION
**Pull request by ragabala**
This commits takes care of adding the following testcases.
1) Test whether a rpm in module is applicablt if appropriate
2) Test Modules that are dependent on other modules
3) Test Whether both modular and non modular rpms can be tested for
applicability
4) Test applicability doesn't work for future RPM versions.

Refer [pulp.plan 4158](https://pulp.plan.io/issues/4158)
Closes #4179

### CODE REVIEW

File: pulp_2_tests/tests/rpm/api_v2/test_content_applicability.py [hunk 1]
```diff
@@ -259,3 +282,204 @@ def test_negative(self):
...
+
+    1. Create a consumer.
+    2. Bind the consumer to the modular repository
+       ``RPM_WITH_MODULES_FEED_URL``.
+    3. Create a consumer profile with:
+       * List of RPMs
+       * List of Modules.
+    4. Regenerate applicability for the consumer.
+    5. Fetch applicability for the consumer. Verify that the packages
+        are eligible for upgrade.
```
Comment (kersommoura) [inline pulp_2_tests/tests/rpm/api_v2/test_content_applicability.py]:
```suggestion
      are eligible for upgrade.
```

File: pulp_2_tests/tests/rpm/api_v2/test_content_applicability.py [hunk 2]
```diff
@@ -259,3 +282,204 @@ def test_negative(self):
...
+    """
+
+    @classmethod
+    def setUpClass(cls):
+        """Create and sync a repository.
+
+        The regular test methods that run after this can create consumers that
+        bind to this repository.
+        """
+        cls.cfg = config.get_config()
```
Comment (kersommoura) [inline pulp_2_tests/tests/rpm/api_v2/test_content_applicability.py]:
```suggestion
        cls.cfg = config.get_config()
        if cfg.pulp_version < Version('2.18'):
           raise unittest.SkipTest('This test requires Pulp 2.18 or newer.')
```

File: pulp_2_tests/tests/rpm/api_v2/test_content_applicability.py [hunk 3]
```diff
@@ -259,3 +282,204 @@ def test_negative(self):
...
+        cls.client = api.Client(cls.cfg, api.json_handler)
+
+    def test_modular_rpm(self):
+        """Verify content is made available if appropriate.
+
+        This Tests the following:
```
Comment (kersommoura) [inline pulp_2_tests/tests/rpm/api_v2/test_content_applicability.py]: How about adding a break line here?

File: pulp_2_tests/tests/rpm/api_v2/test_content_applicability.py [hunk 4]
```diff
@@ -259,3 +282,204 @@ def test_negative(self):
...
+            self.assertEqual(
+                len(applicability[0]['applicability']['modulemd']),
+                1,
+                applicability[0]['applicability']['modulemd'],
+            )
+
+    def test_mixed_rpm(self):
+        """Verify content is made available for both modular/non modular rpms.
+
+        This Tests the following:
```
Comment (kersommoura) [inline pulp_2_tests/tests/rpm/api_v2/test_content_applicability.py]: How about adding a break line here?

File: pulp_2_tests/tests/rpm/api_v2/test_content_applicability.py [hunk 5]
```diff
@@ -259,3 +282,204 @@ def test_negative(self):
...
+        1. Create a consumer profile containing both modular and non modular
+           rpms.
```
Comment (kersommoura) [inline pulp_2_tests/tests/rpm/api_v2/test_content_applicability.py]:
```suggestion
           RPMS.
```

File: pulp_2_tests/tests/rpm/api_v2/test_content_applicability.py [hunk 6]
```diff
@@ -259,3 +282,204 @@ def test_negative(self):
...
+            self.assertEqual(
+                len(applicability[0]['applicability']['modulemd']),
+                0,
+                applicability[0]['applicability']['modulemd'],
+            )
+
+    def do_test(self, modules_profile, rpm_profile):
+        """Regenerate and fetch applicability for the given modules and Rpms.
+
+        This method does the following:
```
Comment (kersommoura) [inline pulp_2_tests/tests/rpm/api_v2/test_content_applicability.py]: How about adding a break line here?

File: pulp_2_tests/tests/rpm/api_v2/test_content_applicability.py [hunk 7]
```diff
@@ -259,3 +282,204 @@ def test_negative(self):
...
+        1. Create a Modular Repo in pulp
+        2. Create a consumer and bind them to the modular repo.
+        3. Create consumer profiles for the passed modules and rpms.
```
Comment (kersommoura) [inline pulp_2_tests/tests/rpm/api_v2/test_content_applicability.py]:
```suggestion
        3. Create consumer profiles for the passed modules and RPMS.
```

File: pulp_2_tests/tests/rpm/api_v2/test_content_applicability.py [hunk 8]
```diff
@@ -259,3 +282,204 @@ def test_negative(self):
...
+
+        # Regenerate applicability.
+        self.client.post(CONSUMERS_ACTIONS_CONTENT_REGENERATE_APPLICABILITY_PATH, {
+            'consumer_criteria': {
+                'filters': {'id': {'$in': [consumer['consumer']['id']]}}
+            }
+        })
+
+        # Fetch applicability.
+        applicability = self.client.post(CONSUMERS_CONTENT_APPLICABILITY_PATH, {
```
Comment (kersommoura) [inline pulp_2_tests/tests/rpm/api_v2/test_content_applicability.py]:
```suggestion
        return self.client.post(CONSUMERS_CONTENT_APPLICABILITY_PATH, {
```

File: pulp_2_tests/tests/rpm/api_v2/test_content_applicability.py [hunk 9]
```diff
@@ -259,3 +282,204 @@ def test_negative(self):
...
+            'criteria': {
+                'filters': {'id': {'$in': [consumer['consumer']['id']]}}
+            },
+        })
+
+        return applicability
```
Comment (kersommoura) [inline pulp_2_tests/tests/rpm/api_v2/test_content_applicability.py]:
```suggestion

```

File: pulp_2_tests/constants.py [hunk 10]
```diff
@@ -171,6 +181,26 @@
 .. _modules.yaml: https://github.com/PulpQE/pulp-fixtures/blob/master/rpm/assets/modules.yaml
 """
 
+MODULE_RPM_DATA = MappingProxyType({
```
Comment (kersommoura) [inline pulp_2_tests/constants.py]: What about?
```suggestion
MODULE_ARTIFACT_RPM_DATA = MappingProxyType({
```

File: pulp_2_tests/constants.py [hunk 11]
```diff
@@ -171,6 +181,26 @@
...
+    'name': 'walrus',
+    'version': '5.21',
+    'release': 'livebeef',
+    'arch': 'noarch',
+    'epoch': '0',
+    'vendor': 'GPLv2'
+})
+"""Details of the RPM file associated with ``MODULE_ERRATA_RPM_DATA``"""
+
+MODULE_RPM_DATA_2 = MappingProxyType({
```
Comment (kersommoura) [inline pulp_2_tests/constants.py]: What about?
```suggestion
MODULE_ARTIFACT_RPM_DATA_2 = MappingProxyType({
```

**Comment by ragabala** [comment]
Hi @kersommoura,

Thanks for those review comments :smile: 
Have made the changes. Please let me know if this works.

**Comment by kersommoura** [comment]
@rochacbruno, @bherrin3 any other comment?

File: pulp_2_tests/tests/rpm/api_v2/test_content_applicability.py [hunk 12]
```diff
@@ -259,3 +283,206 @@ def test_negative(self):
...
+        if cls.cfg.pulp_version < Version('2.18'):
+            raise unittest.SkipTest('This test requires Pulp 2.18 or newer.')
+        cls.client = api.Client(cls.cfg, api.json_handler)
+
+    def test_modular_rpm(self):
+        """Verify content is made available if appropriate.
+
+        This Tests the following:
+
+        1. Create a consumer profile with RPM less version than the
```
Comment (bherrin3) [inline pulp_2_tests/tests/rpm/api_v2/test_content_applicability.py]:
```suggestion
        1. Create a consumer profile with an RPM version less than the
```
tl;dr - You choose. I gave you what I suggest above after asking other experts. Explanation below.  :) 
---
While `version` may be implied, it reads smoother to see the `RPM version` of the consumer profile first. Then, comparing to the module make more sense in the weird way English works. Else, it reads "a consumer profile with an RPM-less version" or an RPM without a version.

Changing `a RPM` to `an RPM` is a contentious change, but preferred by American English speakers. Typically, the indefinite article, `a` or `an`, if preceding a vowel sound is `an` or a consonant sound with `a`. However, my trusted native English speaking friends (who have BAs in Language Arts) tell me that `RPM` has a vowel sound... and "sounds better with `an`". In short, it depends if it `sounds` like a vowel or consonant, not if it IS a vowel or consonant. 

>  R is pronounced "aaaarr" with an open (vowel) sound. The same way you drive `an STi`. S is pronounced "ess" S and R are both consonants, but pronounced with open vowel sounds. You may get `a UTI`. U is a vowel, but pronounced with a Y consonant sound, therefore `a` instead of `an`.

I always hated the a/an rule, so I thought I would give me input above and bother a Bachelor's of Arts graduate.

**Comment by ragabala** [comment]
Thanks, @bherrin3, Good catch, and good explanation.

Have made the changes.
